src/extension/gogoduck/src/main/java/au/org/emii/gogoduck/worker/GoGoDuck.java

```diff
@@ -0,0 +1,511 @@
+
+            List<String> command = new ArrayList<String>();
+            command.add(ncksPath);
+            command.add("-a");
+            command.add("-4");
+            command.add("-O");
+            command.addAll(ncksSubsetParameters);
+            command.addAll(ncksExtraParameters);
+
+            command.add(file.getPath());
+            command.add(tmpFile.getPath());
+
+            logger.info(String.format("Applying subset '%s' to '%s'", ncksSubsetParameters, file.toPath()));
+            execute(command);
+
+            Files.delete(file.toPath());
+            Files.move(tmpFile.toPath(), file.toPath());
+        }
+        catch (Exception e) {
+            throw new GoGoDuckException(String.format("Could not apply subset to file '%s': '%s'", file.getPath(), e.getMessage()));
+        }
+    }
+
+    private static void applySubsetSingleFileNative(File file, GoGoDuckModule module) {
+        // TODO implement!
+    }
+
+    private static class ncksRunnable implements Runnable {
+        private static final Logger logger = LoggerFactory.getLogger(GoGoDuck.class);
+
+        private String name;
+        private Deque<File> workQueue;
+        private GoGoDuckModule module = null;
+        private GoGoDuck gogoduck = null;
+        private ProgressListener progressListener = null;
+
+        ncksRunnable(String name, Deque<File> workQueue, GoGoDuckModule module, GoGoDuck gogoduck) {
+            this.name = name;
+            this.workQueue = workQueue;
+            this.module = module;
+            this.gogoduck = gogoduck;
+        }
+
+        public void run() {
+            try {
+                File file = null;
+                while ((file = workQueue.pop()) != null) {
+                    if (gogoduck.isCancelled()) {
+                        gogoduck.userLog.log("Job was cancelled");
+                        logger.warn("Cancelled by progress listener.");
+                        return;
+                    }
+                    gogoduck.userLog.log(String.format("Processing file '%s'", file.toPath().getFileName()));
+                    applySubsetSingleFileNcks(file, module);
+                }
+            }
+            catch (NoSuchElementException e) {
+                logger.info(String.format("Thread %s finished successfully", name));
+            }
+        }
+    }
+
+    private void applySubsetMultiThread(Path tmpDir, GoGoDuckModule module, int threadCount) throws GoGoDuckException {
+        userLog.log(String.format("Applying subset '%s'", module.getSubsetParameters().toString()));
+        logger.info(String.format("Applying subset on directory '%s'", tmpDir));
+
+        File[] directoryListing = tmpDir.toFile().listFiles();
+        logger.info(String.format("Subset for operation is '%s'", module.getSubsetParameters()));
+
+        Deque<File> workQueue = new ArrayDeque<File>();
+        for (File file : directoryListing) {
+            workQueue.push(file);
+        }
+
+        try {
```

Comment:
Have a look at [ExecutorService usage examples](http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ExecutorService.html).  You may be able to simplify this code a bit, e.g.:

```java

...
// note - change NcksRunnable to operate on a single file, rather than pulling files off a queue, e.g. constructor such as:
ncksRunnable(String name, File file, GoGoDuckModule module, GoGoDuck gogoduck)
...

pool = Executors.newFixedThreadPool(threadCount);

for (File file : directoryListing) {
  pool.execute(new NcksRunnable(..., file, ...);
}

pool.awaitTermination(...);
```

I *think* this code is equivalent - but hopefully you get the idea.